### PR TITLE
Fixes Issue #60

### PR DESCRIPTION
**Pull request by MeTaNoV**
Fixes [Issue #60](https://github.com/PolymerElements/iron-input/issues/60)

### CODE REVIEW

**Comment by notwaldorf** [comment]
As mentioned in the PR, we can't rename the property (because it would break anyone that's currently using it), but would you mind adding a giant explanation comment to the property, explaining this? Perhaps that `allowedPattern` isn't a regex, it's a set of characters that are allowed, and things like `{a}` won't work?

File: test/iron-input.html [hunk 1]
```diff
@@ -138,6 +163,27 @@
         assert.equal(input.bindValue, '123');
       });
 
+      test('prevent invalid input works in _onInput and does not use pattern', function() {
```
Comment (notwaldorf) [inline test/iron-input.html]: I think "prevent-invalid-input works" is a good enough test name (the test isn't using `pattern` anywhere, so the test name is confusing outside of this PR)

File: test/iron-input.html [hunk 2]
```diff
@@ -138,6 +163,27 @@
...
+        var input = fixture('prevent-invalid-input-with-pattern');
+        input.value = '123';
+        input._onInput();
+        assert.equal(input.bindValue, '123');
+        input.validate();
```
Comment (notwaldorf) [inline test/iron-input.html]: This test is testing that `validate` works, not `prevent-invalid-input`. As I mentioned in the original issue, `validate` uses `pattern` regardless of what `prevent-invalid-input` does.

Maybe this test should be renamed to `inputs can be validated`? Which is still a valid test :)

File: test/iron-input.html [hunk 3]
```diff
@@ -120,6 +132,19 @@
         assert.ok(!scope.$.input.value, 'value is falsy');
       });
 
+      test('constraints api without validator', function() {
```
Comment (notwaldorf) [inline test/iron-input.html]: I would rename this test to "can validate using a complex regex"

File: test/iron-input.html [hunk 4]
```diff
@@ -120,6 +132,19 @@
...
+        var input = fixture('no-validator');
+        input.value = '123';
+        input.validate();
+        assert.isTrue(input.invalid, 'input is invalid');
```
Comment (notwaldorf) [inline test/iron-input.html]: I would also add a case when the input is valid :)

**Comment by MeTaNoV** [comment]
I'll rework the comment tomorrow, bed time in Europe! :)

File: test/iron-input.html (hunk 4 above)
Comment (MeTaNoV) [inline test/iron-input.html]: I'll do that too! Thx for the review!

File: test/iron-input.html (hunk 1 above)
Comment (MeTaNoV) [inline test/iron-input.html]: do that too! Thx for the review!

**Comment by notwaldorf** [comment]
Sure thing, no rush :)

File: test/iron-input.html (hunk 2 above)
Comment (MeTaNoV) [inline test/iron-input.html]: It does both test that `validate` is working with `pattern` and that `pattern` is not interfering with `prevent-invalid-input`. Let's go for `inputs can be validated` :)

File: test/iron-input.html (hunk 3 above)
Comment (MeTaNoV) [inline test/iron-input.html]: Yo!

**Comment by MeTaNoV** [comment]
:grrrrr: , I did the modification from work and my git was configured differently...

**Comment by MeTaNoV** [comment]
@notwaldorf how to fix the cla in this case? :)

**Comment by MeTaNoV** [comment]
Ignore the CLA bot.

**Comment by notwaldorf** [comment]
As I mentioned in https://github.com/PolymerElements/paper-input/issues/264#issuecomment-161112988, it seems that if I do this:

```
<paper-input label="only type numbers" auto-validate allowed-pattern="[0-9]" error-message="numbers only!"></paper-input>
```

It automatically sets `prevent-invalid-pattern`, which I don't necessarily want.

**Comment by MeTaNoV** [comment]
Jup, but it does not affect this PR basically. Pattern shouldn't be used to prevent invalid input in the current state. How to tell Mr Bot that CLA is OK?
I took your comment into account, it should be ok to accept?

**Comment by notwaldorf** [comment]
@MeTaNoV Ignoring the CLA bot, this change is a breaking change, since it will be impossible to do what today you can do with: 
```
<paper-input label="only type numbers" auto-validate allowed-pattern="[0-9]" error-message="numbers only!"></paper-input>
```

Which is a good feature to have. Since this change will break this existing functionality, it cannot be merged. I think https://github.com/PolymerElements/iron-input/pull/43 introduced automatically setting `prevent-invalid-input`, which doesn't work with your change. For this to land, that change would need to be undone.

**Comment by MeTaNoV** [comment]
of course you can, setting `allowed-pattern` is independant from `pattern` I added test, and no test broke!

**Comment by notwaldorf** [comment]
But as of #43, setting `allowed-pattern` automatically forces `prevent-invalid-input`, which means this PR (that won't allow me to set `pattern` and have that be considered in the auto-validation) will break current functionality.

Basically: for this PR to land, #43 needs to be reverted (by you adding the opposite code, basically, so that we can keep track of why it was changed), or, if #43 stays, this PR cannot land.

**Comment by MeTaNoV** [comment]
try to do a test setting in one test `allowed-pattern` and the other setting `pattern` and you will see that it work perfectly, I don't see any problem. Setting `pattern` won't force `prevent-invalid-input` hopefully and it should not.

**Comment by notwaldorf** [comment]
So then I don't understand your original comment from the other thread, that I shouldn't do
```
<paper-input label="only type numbers" auto-validate pattern="[0-9]*" error-message="numbers only!"></paper-input>
```

What I want is to be able to type all the possible characters, but have the `paper-input` display whether or not what I've typed is invalid

**Comment by MeTaNoV** [comment]
this was a mistake! ;)

**Comment by notwaldorf** [comment]
What was a mistake, your comment, or my sample code? My main goal is trying to understand if this PR will break any existing behaviour.

**Comment by MeTaNoV** [comment]
my comment on the other thread, and it does not affect this PR! :)

**Comment by MeTaNoV** [comment]
sorry for the puzzling effect!

**Comment by MeTaNoV** [comment]
for this one CI from Travis is all OK, I just did the commit from work with other git ID, that's why GoogleBot is suspicious...

**Comment by notwaldorf** [comment]
@MeTaNoV Yup, just re-running the tests, since this has been open for a while, and if it's fine, I'll merge it. Thanks for sticking through it!

**Comment by MeTaNoV** [comment]
thanks for the follow up and sorry again for the confusion! I know it's hard to switch context all the time!

**Comment by notwaldorf** [comment]
No worries!

**Comment by MeTaNoV** [comment]
+1

**Pull request merged**